scanpy/tests/test_clustering.py

```diff
@@ -20,6 +21,52 @@ def test_leiden_basic(adata_neighbors):
     sc.tl.leiden(adata_neighbors)
 
 
+def test_clustering_subset(adata_neighbors):
+    methods = (
+        (sc.tl.louvain, 'louvain'),
+        (sc.tl.leiden, 'leiden'),
+    )
+    for clustering, key in methods:
+        print('Clustering method ' + str(key))
+        clustering(adata_neighbors, key_added=key)
+        print(adata_neighbors)
+        print(adata_neighbors.obs[key].value_counts())
+
+        for c in np.unique(adata_neighbors.obs[key]):
+            print('Analyzing cluster ' + str(c))
+            cells_in_c = (adata_neighbors.obs[key] == c)
+            ncells_in_c = adata_neighbors.obs[key].value_counts().loc[c]
+            key_sub = str(key) + '_sub'
+            clustering(adata_neighbors, restrict_to=(key, [c]),
+                key_added=key_sub) 
+
+            # Old category should not be present
+            assert (c not in
+                adata_neighbors.obs.loc[:, key_sub].value_counts().index)
+
+            # Cells in the original category are assigned only to new categories
+            print('Categories')
+            cat_counts = adata_neighbors.obs.loc[cells_in_c, key_sub].value_counts()
+            print(cat_counts)
+
+            ## Only cells of the original cluster have been assigned to
+            ## new categories
+            assert cat_counts.sum() == ncells_in_c
+
+            print('Non-zero categories')
+            nonzero_cat = (cat_counts > 0)
+            print(nonzero_cat)
+            nonzero_cat = (adata_neighbors.obs[key_sub].cat.categories
+                .to_series()[nonzero_cat].index)
+            print(nonzero_cat)
```

Comment:
You assign `nonzero_cat` twice. The first is a boolean array you use for indexing and the second is, too (you do `nonzero_cat & ...`)? I’m confused here.

Reply:
I think those two assignments could be simplified to:

```python
nonzero_cat = cat_counts[cat_counts != 0].index
```

Reply:
@flying-sheep, `pd.Index`s have a lot of set-like features, so `&`  here is an intersection operator.

Reply:
I partially rewrote that part (it was confusing, sorry). And implemented @ivirshup 's suggestion. Let me know how it looks.